### PR TITLE
Add in --duplex flag to run HTTP and GRPC servers on the same port

### PR DESCRIPTION
**Pull request by priyawadhwa**
also adds in a unit test for this. we can optionally start trying this out in staging, and if it works well we can consider deprecating the previous two-port solution.

Signed-off-by: Priya Wadhwa <priya@chainguard.dev>

#### Release Note
Add in --duplex flag to run HTTP and GRPC servers on the same port

cc @vaikas

### CODE REVIEW

**Comment by haydentherapper** [comment]
Thanks! Let's hold this until the new year, and we can cut a new release at the same time.

**Comment by haydentherapper** [comment]
@priyawadhwa Is there any change in behavior between serving gRPC on its own port and duplex? Any differences for metrics?

**Comment by priyawadhwa** [comment]
There shouldn't be!

File: cmd/app/serve.go
Comment (vaikas) [inline cmd/app/serve.go:103]: Should we also mark grpc-host/grpc-port deprecated in their flag descriptions? Then maybe later on we can use something like:
https://pkg.go.dev/go.etcd.io/etcd/pkg/flags#DeprecatedFlag

**Review by vaikas — COMMENTED** [review]
Awesomesauce!

File: cmd/app/serve.go
Comment (priyawadhwa) [reply cmd/app/serve.go:103]: Added! I also added in a warning to suggest to users that they switch to using the --duplex flag if it isn't set.

Comment (vaikas) [reply cmd/app/serve.go:103]: Thanks!, that's a great idea. wonder if we should do something about the http-host / http-port aliases as well, since with duplex they are kind of wonky? Up to you, but now that I think about it, in the future they too should be deprecated, removed and only use host/port?

Comment (priyawadhwa) [reply cmd/app/serve.go:103]: I think it makes sense to remove the aliases! They were added in back in April (https://github.com/sigstore/fulcio/pull/472) so I think it's been enough time since then. I guess whenever we actually remove grpc-port grpc-host we can remove http-port and http-host as well.

**Review by vaikas — APPROVED** [review]

**Comment by haydentherapper** [comment]
Thanks y’all, I’ll test out these changes after the holidays!

File: cmd/app/serve.go
Comment (priyawadhwa) [reply cmd/app/serve.go:99]: If we eventually plan to remove them (which I think makes the most sense) then it's best to deprecate them sooner. IIUC we'll still maintain support for at least a few releases or a few months (whatever the policy comes up with suggests).

File: cmd/app/serve.go
Comment (priyawadhwa) [reply cmd/app/serve.go:103]: Hmm it would work, but as a user I don't find it very intuitive. As a user I'd expect that setting `--port` and `--grpc-port` to the same value would cause an error.  I'd still prefer to deprecate the grpc flags and have one `--port` flag to expose everything on, mostly because I think longterm it'll be cleaner, easier for users to understand, and because I don't see a use case for GRPC-only functionality.

File: cmd/app/serve.go
Comment (haydentherapper) [reply cmd/app/serve.go:99]: I don't think it's a great look to deprecate functionality right after we've introduced it as stable. I know it's a tiny part of the service, but I think we shouldn't begin deprecation of features right after a 1.0 release until we start thinking about a 2.0 release, and right now, I don't see any features in the pipeline to need a 2.0 release.

File: cmd/app/serve.go
Comment (haydentherapper) [reply cmd/app/serve.go:100]: It might affect those who run their own infrastructure.

File: cmd/app/serve.go
Comment (haydentherapper) [reply cmd/app/serve.go:103]: I see the addition of this change as fixing the error of setting the ports to be the same, rather than adding new functionality. I also would prefer to not use `duplex` as a flag name since that's exposing implementation details.

File: cmd/app/serve.go
Comment (haydentherapper) [inline cmd/app/serve.go:353]: Is `host` configurable?

**Comment by priyawadhwa** [comment]
Discussed with @haydentherapper and made the following changes:
- Removed the `--duplex` flag in favor of using duplex if `--grpc-port` and `--port` are set to the same value.
- Removed the deprecation warnings for now, we can add those back in later once we feel duplex is running as expected.

@haydentherapper this is RFAL.

File: cmd/app/serve.go
Comment (priyawadhwa) [reply cmd/app/serve.go:350]: viper.GetString("port")

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:275]: I think `viper.GetInt("port")` works and saves the `Atoi` call

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:280]: same here on `GetInt`

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:287]: I think we're missing a `return` statement after this, since we wouldn't want to go on and restart everything again

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:405]: could we re-use https://github.com/sigstore/fulcio/blob/bc08c5b641033bcd9f604e202c84c8c9fde502dd/cmd/app/http.go#L97 instead of this implementation? The existing one deals with returning the SCT in a response header (which was part of v1 IIUC), and also strips out all GRPC response headers (to hide the internal implementation detail).

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:274]: Should this also test that `host` and `grpc-host` are equal before proceeding?

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:285]: pass `host` down to `StartDuplexServer`

File: cmd/app/serve.go
Comment (bobcallaway) [inline cmd/app/serve.go:395]:
```suggestion
        lis, err := net.Listen("tcp", fmt.Sprintf("%s:%d", host, port))
	if err != nil {
		return fmt.Errorf("creating listener: %w", err)
	}
	if err := d.Serve(ctx, lis); err != nil {
		return fmt.Errorf("duplex server: %w", err)
	}
```

File: cmd/app/serve.go
Comment (priyawadhwa) [reply cmd/app/serve.go]: Thanks! I hadn't noticed that before, it should work as well. I've updated it to use this one!

**Comment by haydentherapper** [comment]
cc @bobcallaway for another look

**Review by bobcallaway — DISMISSED** [review]

**Review by haydentherapper — COMMENTED** [review]
LGTM, just needs to be rebased for go.mod

File: cmd/app/serve.go
Comment (bobcallaway) [reply cmd/app/serve.go]: it was added in arg list in https://github.com/sigstore/fulcio/pull/931/files#diff-1627906dbce532aff8cc81f54790572e5ba04f3e568c88b6ed39a590642dafa4R345

Comment (haydentherapper) [reply cmd/app/serve.go]: Oh whoops that was a comment I wrote awhile ago. This looks good

**Comment by priyawadhwa** [comment]
Rebased! could i please get another lgtm?

**Review by bobcallaway — APPROVED** [review]

**Pull request merged**
